compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java

```diff
@@ -405,15 +403,8 @@ private static void cacheBIR(ModuleContext moduleContext) {
 
         // Can we improve this logic
         ByteArrayOutputStream birContent = new ByteArrayOutputStream();
-        try {
-            byte[] pkgBirBinaryContent = PackageFileWriter.writePackage(
-                    moduleContext.bLangPackage.symbol.birPackageFile);
-            birContent.writeBytes(pkgBirBinaryContent);
-            moduleContext.compilationCache.cacheBir(moduleContext.moduleName(), birContent);
-        } catch (IOException e) {
-            // This path may never be executed
-            throw new RuntimeException("Failed to convert BIR model to a byte array", e);
-        }
+        birContent.writeBytes(moduleContext.bLangPackage.symbol.birPackageFile.pkgBirBinaryContent);
```

Comment:
Looks like the BIR is already converted to binary before executing this line? I think the project API should have control over generating the binary representation.

Reply:
Yes. We do at Birgen. We keep the binary at BPackageSymbol. I think this was done mainly to ensure serialization done only once and reuse binary content in different places ( bir-dump-file flag, bir cache).

Reply:
https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java#L265

Reply:
Can we do the serialization only when it is required? On-demand basically.

Reply:
We can do it. We can cache it the same way (so we do not need to serialize it multiple times). I'll update the PR.